### PR TITLE
3 of 3 Separate Frames Viewer: Handle various frames layouts

### PR DESCRIPTION
**Pull request by goplayoutside3**
~~⚠️ Contains #4455 please review that one first.~~

## Package
lib-classifier

## Linked Issue and/or Talk Post
Related changes to the FEM lab component in PFE: https://github.com/zooniverse/Panoptes-Front-End/pull/6536

## Describe your changes

- Added `multi_image_layout` to the Workflow Configuration store. This is the property that SeparateFramesViewer component now observes to determine which grid to display.

- I added all four layout options as stories to the classifier's storybook.

- There's a resize observer in SeparateFramesContainer that prevents subject images from becoming too tiny by forcing the layout to 1-column if the SeparateFrame component is ever less than 200px in width

- Fixed the "with key zoom" feature in SeparateFrame. When a frame image is focused, a volunteer can used their keyboard to pan and zoom the image.

## How to Review

Review the layouts in the classifier's storybook.

My test flipbook project does have switching views enabled, so you can see the various layouts, and switch between flipbook and separate frames view while viewing this workflow: https://local.zooniverse.org:3000/projects/goplayoutside3/fem-flipbook-viewer/classify/workflow/22532?env=production

(It's lab: https://www.zooniverse.org/lab/19645/workflows/22532?femLab=true)

You can also look at Daily Minor Planet which as a 1-column separate frames view enabled in its flipbook: https://local.zooniverse.org:3000/projects/fulsdavid/the-daily-minor-planet/classify/workflow/22354?env=production

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [x] Tests are passing locally and on Github
- [x] Documentation is up to date and changelog has been updated if appropriate
- [x] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [x] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [x] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [x] Can submit a classification
- [x] Can sign-in and sign-out
- [x] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [x] The PR creator has listed user actions to use when testing the new feature
- [x] Unit tests are included for the new feature
- [x] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

### CODE REVIEW

**Comment by goplayoutside3** [comment]
@seanmiller26 just wanted to tag you here for a review because we started a convo about responsiveness on the FEM lab PR. Let me know what you think about the following UX:

In this separate frames layout PR, I added code that listens for the width of a separate frame component (including its toolbar), and if that component's width is ever less than 200px, then the layout switches to 1-column. There's a screenshot below that outlines the component I'm referring to.

For instance, the responsive switch to 1-column will happen automatically for a mobile device like a phone. And if a project team chooses the 'row' layout for subjects that contain several frames, this code logic will prevent a volunteer with a device like an iPad from seeing tiny subject images. Happy to demo if needed too!

Screenshot 2023-04-24 at 4 46 45 PM

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/components/SeparateFrame/SeparateFrame.js
Comment (mcbouslog) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/components/SeparateFrame/SeparateFrame.js:235]: Noting that zooming with +/- keys doesn't work on Safari, but it's not related to these changes, it also doesn't work on other projects like https://www.zooniverse.org/projects/humphrydavy/davy-notebooks-project .

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/SeparateFramesViewer.js
Comment (mcbouslog) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/SeparateFramesViewer.js:94]: The ViewModeButton is expanding horizontally in a way that changes the cursor from a an arrow to pointer. It doesn't respond to clicks, but unexpected.
ViewModeButton

**Review by mcbouslog — APPROVED** [review]
Looks great! 🎉 

Noted a few things for consideration, but I don't think anything blocking.

If I'm using the multi-frame viewer, i.e. a transcription workflow, I think the multi-image options will show in the workflow editor, but they should not?

Are there plans to allow the separate frames viewer exclusively (without the multi-image viewer)?

**Comment by seanmiller26** [comment]
> if that component's width is ever less than 200px, then the layout switches to 1-column
Does the screenshot show the component at its minimum size? I'm wondering if 200px is too small.

And related - is there a way to wrap the image to the next line without changing to a 1 column view? Basically do we need an in between mode before it switches to 1 column? I could see arguments for both functionalities.

> Are there plans to allow the separate frames viewer exclusively (without the multi-image viewer)?
I second this question. Could be a good discussion to have with the wider team

**Comment by goplayoutside3** [comment]
@seanmiller26 

> is there a way to wrap the image to the next line without changing to a 1 column view?

I don't think we should do this because it makes the four discrete choices in the project builder more messy and unpredictable.

> Does the screenshot show the component at its minimum size? I'm wondering if 200px is too small.

The screenshot is probably not a good way of measuring. I picked 200px as a happy medium so that mobile phones will show a 1-column layout, yet desktop browsers will still display three or four frames horizontally.

The choice in "breakpoint width" needs to be flexible for project teams that might want to use the separate frames viewer for subjects that have multiple frames, but don't care if those frames images are only about 200px in width (like blurry galaxies). But we also don't want volunteers to look at a layout and close the page because classification is impossible.

Do you recommend a different width in pixels or rem before this PR is merged?

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/components/SeparateFrame/SeparateFrame.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/components/SeparateFrame/SeparateFrame.js:235]: Good to know!

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/SeparateFramesViewer.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SeparateFramesViewer/SeparateFramesViewer.js:94]: Good point, I'll fix this styling before merging

**Comment by goplayoutside3** [comment]
@mcbouslog 

> If I'm using the multi-frame viewer, i.e. a transcription workflow, I think the multi-image options will show in the workflow editor, but they should not? Are there plans to allow the separate frames viewer exclusively (without the multi-image viewer)?

This is a fem-lab question for the PFE codebase. My educated guess right now is that transcription projects on FEM usually still use the pfe lab anyway.

There are plans to add a subject viewer selection dropdown to the fem lab. That dropdown will aid in logic for when and when not to show multi-image options that pertain to separate frames view.

In addition, FEM transcription projects that default to the MultiImageViewer component are not at risk of seeing the ViewModeButton because it only exists in the flipbook controls component so far.

**Comment by seanmiller26** [comment]
Could you show what a row of subjects at 200px each looks like? And then each at 320px?

**Comment by goplayoutside3** [comment]
@seanmiller26 what kinda of subject would be most helpful in screenshots to compare 200px to 320px? For instance, a 1:1 square image like Daily Minor Planet, or landscape images like Wildwatch Burrowing Owl? I just don't want to influence a final decision with my arbitrary test project images

**Comment by seanmiller26** [comment]
Lets try it with Wildwatch Burrowing Owl subjects

**Comment by goplayoutside3** [comment]
@seanmiller26 I deployed this branch to the following link along with a temporary resizing tool that displays the width a frame component as you resize the browser window. This link is Wildwatch Burrowing Owl with view mode button force-enabled, and with a 'row' layout. 

https://fe-project-branch.preview.zooniverse.org/projects/sandiegozooglobal/wildwatch-burrowing-owl/classify/workflow/21141

Switch to separate frames view and you should see this:
Screenshot 2023-04-25 at 3 20 52 PM

The red number in pixels indicates the current width of the component outlined in red. Let me know the pixel number at which the subject image seems too small, and I'll change the "breakpoint" to that in this PR before merging to production!

**Comment by seanmiller26** [comment]
Thanks for this, it's really useful to see it in action. I'm still conflicted because 200px seems almost too small to be useful for classifying. But then again 320px barely lets you fit 3 subjects across, even in a wide screen browser window. I know we're using landscape subjects, but you can extrapolate what it would look like with portrait subjects.

I think I'd still vote for the minimum size to be 320px.

This may have opened a can of worms. Once you get to mobile view at 768px it gets really funky, bouncing between rows and columns. Not sure if we want to address this all at this moment or not, though.

**Comment by goplayoutside3** [comment]
Sounds good! I'll increase the breakpoint and as more flipbook project use FEM, we can always adjust in the future.

**Pull request merged**
